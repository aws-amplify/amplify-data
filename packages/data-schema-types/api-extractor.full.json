{
  "extends": "../../api-extractor.full.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/index.internal.d.ts"
}
