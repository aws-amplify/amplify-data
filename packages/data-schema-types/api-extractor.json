{
  "extends": "../../api-extractor.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/index.d.ts"
}
