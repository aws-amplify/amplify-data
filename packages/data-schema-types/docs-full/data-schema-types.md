<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema-types](./data-schema-types.md)

## data-schema-types package

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[DerivedApiDefinition](./data-schema-types.derivedapidefinition.md)


</td><td>


</td></tr>
<tr><td>

[GraphQLFormattedError](./data-schema-types.graphqlformattederror.md)


</td><td>

See: https://spec.graphql.org/draft/\#sec-Errors


</td></tr>
<tr><td>

[SourceLocation](./data-schema-types.sourcelocation.md)


</td><td>

Represents a location in a Source.


</td></tr>
</tbody></table>

## Variables

<table><thead><tr><th>

Variable


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[\_\_modelMeta\_\_](./data-schema-types.__modelmeta__.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[AuthMode](./data-schema-types.authmode.md)


</td><td>


</td></tr>
<tr><td>

[BackendSecret](./data-schema-types.backendsecret.md)


</td><td>

Importing the full objects from @<!-- -->aws-amplify/plugin-types more than doubles dev env runtime. This type replacement will contain the content for config without the negative side-effects. We may need to re-approach if customers interact with these programmatically to avoid forcing narrowing.


</td></tr>
<tr><td>

[BooleanFilters](./data-schema-types.booleanfilters.md)


</td><td>


</td></tr>
<tr><td>

[ContextType](./data-schema-types.contexttype.md)


</td><td>


</td></tr>
<tr><td>

[CreateModelInput](./data-schema-types.createmodelinput.md)


</td><td>

All identifiers and fields used to create a model


</td></tr>
<tr><td>

[CustomHeaders](./data-schema-types.customheaders.md)


</td><td>

Custom headers that can be passed either to the client or to individual model operations, either as a static object or a function that returns a promise.


</td></tr>
<tr><td>

[CustomMutations](./data-schema-types.custommutations.md)


</td><td>


</td></tr>
<tr><td>

[CustomOperationFnParams](./data-schema-types.customoperationfnparams.md)


</td><td>

Generates Custom Operations function params based on whether .arguments() were specified in the schema builder


</td></tr>
<tr><td>

[CustomOperationMethodOptions](./data-schema-types.customoperationmethodoptions.md)


</td><td>


</td></tr>
<tr><td>

[CustomOperations](./data-schema-types.customoperations.md)


</td><td>


</td></tr>
<tr><td>

[CustomQueries](./data-schema-types.customqueries.md)


</td><td>


</td></tr>
<tr><td>

[CustomSelectionSetReturnValue](./data-schema-types.customselectionsetreturnvalue.md)


</td><td>

Generates flattened, readonly return type using specified custom sel. set


</td></tr>
<tr><td>

[CustomSqlDataSourceStrategy](./data-schema-types.customsqldatasourcestrategy.md)


</td><td>


</td></tr>
<tr><td>

[CustomSubscriptions](./data-schema-types.customsubscriptions.md)


</td><td>


</td></tr>
<tr><td>

[DataSourceConfiguration](./data-schema-types.datasourceconfiguration.md)


</td><td>


</td></tr>
<tr><td>

[DatasourceEngine](./data-schema-types.datasourceengine.md)


</td><td>


</td></tr>
<tr><td>

[Debug](./data-schema-types.debug.md)


</td><td>


</td></tr>
<tr><td>

[DeepPickFromPath](./data-schema-types.deeppickfrompath.md)


</td><td>

Picks object properties that match provided dot-separated Path


</td></tr>
<tr><td>

[DeepReadOnlyObject](./data-schema-types.deepreadonlyobject.md)


</td><td>


</td></tr>
<tr><td>

[DefineFunction](./data-schema-types.definefunction.md)


</td><td>


</td></tr>
<tr><td>

[DerivedCombinedSchema](./data-schema-types.derivedcombinedschema.md)


</td><td>


</td></tr>
<tr><td>

[DerivedModelSchema](./data-schema-types.derivedmodelschema.md)


</td><td>


</td></tr>
<tr><td>

[EnumTypes](./data-schema-types.enumtypes.md)


</td><td>

The utility type that is used to infer the type (interface) of the generated `client.enums` property.


</td></tr>
<tr><td>

[Equal](./data-schema-types.equal.md)


</td><td>


</td></tr>
<tr><td>

[ExcludeEmpty](./data-schema-types.excludeempty.md)


</td><td>


</td></tr>
<tr><td>

[Expect](./data-schema-types.expect.md)


</td><td>


</td></tr>
<tr><td>

[ExpectFalse](./data-schema-types.expectfalse.md)


</td><td>


</td></tr>
<tr><td>

[ExpectTrue](./data-schema-types.expecttrue.md)


</td><td>


</td></tr>
<tr><td>

[ExtractModelMeta](./data-schema-types.extractmodelmeta.md)


</td><td>


</td></tr>
<tr><td>

[FunctionSchemaAccess](./data-schema-types.functionschemaaccess.md)


</td><td>


</td></tr>
<tr><td>

[HasKey](./data-schema-types.haskey.md)


</td><td>


</td></tr>
<tr><td>

[IfEquals](./data-schema-types.ifequals.md)


</td><td>


</td></tr>
<tr><td>

[IndexQueryMethodsFromIR](./data-schema-types.indexquerymethodsfromir.md)


</td><td>


</td></tr>
<tr><td>

[IndexQueryMethodSignature](./data-schema-types.indexquerymethodsignature.md)


</td><td>


</td></tr>
<tr><td>

[IsAny](./data-schema-types.isany.md)


</td><td>


</td></tr>
<tr><td>

[IsEmptyStringOrNever](./data-schema-types.isemptystringornever.md)


</td><td>


</td></tr>
<tr><td>

[IsFalse](./data-schema-types.isfalse.md)


</td><td>


</td></tr>
<tr><td>

[IsTrue](./data-schema-types.istrue.md)


</td><td>


</td></tr>
<tr><td>

[JsResolver](./data-schema-types.jsresolver.md)


</td><td>


</td></tr>
<tr><td>

[JsResolverEntry](./data-schema-types.jsresolverentry.md)


</td><td>


</td></tr>
<tr><td>

[LambdaFunctionDefinition](./data-schema-types.lambdafunctiondefinition.md)


</td><td>


</td></tr>
<tr><td>

[LazyLoader](./data-schema-types.lazyloader.md)


</td><td>


</td></tr>
<tr><td>

[ListReturnValue](./data-schema-types.listreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[LogicalFilters](./data-schema-types.logicalfilters.md)


</td><td>


</td></tr>
<tr><td>

[Model](./data-schema-types.model.md)


</td><td>


</td></tr>
<tr><td>

[ModelFilter](./data-schema-types.modelfilter.md)


</td><td>


</td></tr>
<tr><td>

[ModelIdentifier](./data-schema-types.modelidentifier.md)


</td><td>


</td></tr>
<tr><td>

[ModelMetaShape](./data-schema-types.modelmetashape.md)


</td><td>


</td></tr>
<tr><td>

[ModelPath](./data-schema-types.modelpath.md)


</td><td>

Generates custom selection set type with up to 6 levels of nested fields


</td></tr>
<tr><td>

[ModelSortDirection](./data-schema-types.modelsortdirection.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypes](./data-schema-types.modeltypes.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypesClient](./data-schema-types.modeltypesclient.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypesSSRCookies](./data-schema-types.modeltypesssrcookies.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypesSSRRequest](./data-schema-types.modeltypesssrrequest.md)


</td><td>


</td></tr>
<tr><td>

[MutationInput](./data-schema-types.mutationinput.md)


</td><td>

All required fields and relational fields, exclude readonly fields


</td></tr>
<tr><td>

[NonRelationalFields](./data-schema-types.nonrelationalfields.md)


</td><td>

Currently this omits any object-type fields. Update this when we add custom types/enums.


</td></tr>
<tr><td>

[NotAny](./data-schema-types.notany.md)


</td><td>


</td></tr>
<tr><td>

[NotEqual](./data-schema-types.notequal.md)


</td><td>


</td></tr>
<tr><td>

[NumericFilter](./data-schema-types.numericfilter.md)


</td><td>


</td></tr>
<tr><td>

[ObjectIsNonEmpty](./data-schema-types.objectisnonempty.md)


</td><td>


</td></tr>
<tr><td>

[ObservedReturnValue](./data-schema-types.observedreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[ObserveQueryReturnValue](./data-schema-types.observequeryreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[PathEntry](./data-schema-types.pathentry.md)


</td><td>

references IAmplifyGraphqlDefinition from: https://github.com/aws-amplify/amplify-category-api/blob/4c0ea253a0bae51f775383929ba4748593185bc1/packages/amplify-graphql-api-construct/src/types.ts\#L491-L503

function slots is any'd for now. Will add actual type when we add support for this feature


</td></tr>
<tr><td>

[Prettify](./data-schema-types.prettify.md)


</td><td>


</td></tr>
<tr><td>

[RequestOptions](./data-schema-types.requestoptions.md)


</td><td>

Request options that are passed to custom header functions. `method` and `headers` are not included in custom header functions passed to subscriptions.


</td></tr>
<tr><td>

[ResolvedModel](./data-schema-types.resolvedmodel.md)


</td><td>

Flattens model instance type and unwraps async functions into resolved GraphQL shape

This type is used for generating the base shape for custom selection set input and its return value Uses same pattern as above to limit recursion depth to maximum usable for selection set.


</td></tr>
<tr><td>

[RestoreArrays](./data-schema-types.restorearrays.md)


</td><td>

This mapped type traverses the SelectionSetReturnValue result and the original FlatModel, restoring array types that were flattened in DeepPickFromPath


</td></tr>
<tr><td>

[ReturnValue](./data-schema-types.returnvalue.md)


</td><td>

Selection set-aware CRUDL operation return value type


</td></tr>
<tr><td>

[SchemaConfiguration](./data-schema-types.schemaconfiguration.md)


</td><td>


</td></tr>
<tr><td>

[SecondaryIndexIrShape](./data-schema-types.secondaryindexirshape.md)


</td><td>

SecondaryIndex index types and query methods


</td></tr>
<tr><td>

[SelectionSet](./data-schema-types.selectionset.md)


</td><td>


</td></tr>
<tr><td>

[SetTypeSubArg](./data-schema-types.settypesubarg.md)


</td><td>

Replaces the value of a key in a complex generic type param


</td></tr>
<tr><td>

[SingularReturnValue](./data-schema-types.singularreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[SizeFilter](./data-schema-types.sizefilter.md)


</td><td>

Filter options that can be used on fields where size checks are supported.


</td></tr>
<tr><td>

[SqlStatementFolderEntry](./data-schema-types.sqlstatementfolderentry.md)


</td><td>


</td></tr>
<tr><td>

[StringFilter](./data-schema-types.stringfilter.md)


</td><td>

Filters options that can be used on string-like fields.


</td></tr>
<tr><td>

[SubnetAZ](./data-schema-types.subnetaz.md)


</td><td>


</td></tr>
<tr><td>

[UnionToIntersection](./data-schema-types.uniontointersection.md)


</td><td>


</td></tr>
<tr><td>

[UnwrapArray](./data-schema-types.unwraparray.md)


</td><td>


</td></tr>
<tr><td>

[VpcConfig](./data-schema-types.vpcconfig.md)


</td><td>


</td></tr>
<tr><td>

[WritableKeys](./data-schema-types.writablekeys.md)


</td><td>


</td></tr>
</tbody></table>
