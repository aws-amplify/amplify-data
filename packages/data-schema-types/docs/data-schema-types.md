<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema-types](./data-schema-types.md)

## data-schema-types package

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[DerivedApiDefinition](./data-schema-types.derivedapidefinition.md)


</td><td>

references IAmplifyGraphqlDefinition from: https://github.com/aws-amplify/amplify-category-api/blob/4c0ea253a0bae51f775383929ba4748593185bc1/packages/amplify-graphql-api-construct/src/types.ts\#L491-L503

function slots is any'd for now. Will add actual type when we add support for this feature


</td></tr>
<tr><td>

[GraphQLFormattedError](./data-schema-types.graphqlformattederror.md)


</td><td>

See: https://spec.graphql.org/draft/\#sec-Errors


</td></tr>
<tr><td>

[SourceLocation](./data-schema-types.sourcelocation.md)


</td><td>

Represents a location in a Source.


</td></tr>
</tbody></table>

## Variables

<table><thead><tr><th>

Variable


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[\_\_modelMeta\_\_](./data-schema-types.__modelmeta__.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[AuthMode](./data-schema-types.authmode.md)


</td><td>


</td></tr>
<tr><td>

[CustomHeaders](./data-schema-types.customheaders.md)


</td><td>

Custom headers that can be passed either to the client or to individual model operations, either as a static object or a function that returns a promise.


</td></tr>
<tr><td>

[CustomMutations](./data-schema-types.custommutations.md)


</td><td>


</td></tr>
<tr><td>

[CustomOperations](./data-schema-types.customoperations.md)


</td><td>


</td></tr>
<tr><td>

[CustomQueries](./data-schema-types.customqueries.md)


</td><td>


</td></tr>
<tr><td>

[CustomSqlDataSourceStrategy](./data-schema-types.customsqldatasourcestrategy.md)


</td><td>


</td></tr>
<tr><td>

[CustomSubscriptions](./data-schema-types.customsubscriptions.md)


</td><td>


</td></tr>
<tr><td>

[DataSourceConfiguration](./data-schema-types.datasourceconfiguration.md)


</td><td>


</td></tr>
<tr><td>

[DatasourceEngine](./data-schema-types.datasourceengine.md)


</td><td>


</td></tr>
<tr><td>

[Debug](./data-schema-types.debug.md)


</td><td>


</td></tr>
<tr><td>

[DeepReadOnlyObject](./data-schema-types.deepreadonlyobject.md)


</td><td>


</td></tr>
<tr><td>

[DefineFunction](./data-schema-types.definefunction.md)


</td><td>


</td></tr>
<tr><td>

[DerivedCombinedSchema](./data-schema-types.derivedcombinedschema.md)


</td><td>


</td></tr>
<tr><td>

[DerivedModelSchema](./data-schema-types.derivedmodelschema.md)


</td><td>


</td></tr>
<tr><td>

[EnumTypes](./data-schema-types.enumtypes.md)


</td><td>

The utility type that is used to infer the type (interface) of the generated `client.enums` property.


</td></tr>
<tr><td>

[Equal](./data-schema-types.equal.md)


</td><td>


</td></tr>
<tr><td>

[ExcludeEmpty](./data-schema-types.excludeempty.md)


</td><td>


</td></tr>
<tr><td>

[Expect](./data-schema-types.expect.md)


</td><td>


</td></tr>
<tr><td>

[ExpectFalse](./data-schema-types.expectfalse.md)


</td><td>


</td></tr>
<tr><td>

[ExpectTrue](./data-schema-types.expecttrue.md)


</td><td>


</td></tr>
<tr><td>

[ExtractModelMeta](./data-schema-types.extractmodelmeta.md)


</td><td>


</td></tr>
<tr><td>

[FunctionSchemaAccess](./data-schema-types.functionschemaaccess.md)


</td><td>


</td></tr>
<tr><td>

[HasKey](./data-schema-types.haskey.md)


</td><td>


</td></tr>
<tr><td>

[IsAny](./data-schema-types.isany.md)


</td><td>


</td></tr>
<tr><td>

[IsEmptyStringOrNever](./data-schema-types.isemptystringornever.md)


</td><td>


</td></tr>
<tr><td>

[IsFalse](./data-schema-types.isfalse.md)


</td><td>


</td></tr>
<tr><td>

[IsTrue](./data-schema-types.istrue.md)


</td><td>


</td></tr>
<tr><td>

[JsResolver](./data-schema-types.jsresolver.md)


</td><td>


</td></tr>
<tr><td>

[JsResolverEntry](./data-schema-types.jsresolverentry.md)


</td><td>


</td></tr>
<tr><td>

[LambdaFunctionDefinition](./data-schema-types.lambdafunctiondefinition.md)


</td><td>


</td></tr>
<tr><td>

[LazyLoader](./data-schema-types.lazyloader.md)


</td><td>


</td></tr>
<tr><td>

[ListReturnValue](./data-schema-types.listreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[ModelPath](./data-schema-types.modelpath.md)


</td><td>

Generates custom selection set type with up to 6 levels of nested fields


</td></tr>
<tr><td>

[ModelSortDirection](./data-schema-types.modelsortdirection.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypes](./data-schema-types.modeltypes.md)


</td><td>


</td></tr>
<tr><td>

[NotAny](./data-schema-types.notany.md)


</td><td>


</td></tr>
<tr><td>

[NotEqual](./data-schema-types.notequal.md)


</td><td>


</td></tr>
<tr><td>

[ObjectIsNonEmpty](./data-schema-types.objectisnonempty.md)


</td><td>


</td></tr>
<tr><td>

[ObservedReturnValue](./data-schema-types.observedreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[ObserveQueryReturnValue](./data-schema-types.observequeryreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[Prettify](./data-schema-types.prettify.md)


</td><td>


</td></tr>
<tr><td>

[RequestOptions](./data-schema-types.requestoptions.md)


</td><td>

Request options that are passed to custom header functions. `method` and `headers` are not included in custom header functions passed to subscriptions.


</td></tr>
<tr><td>

[SchemaConfiguration](./data-schema-types.schemaconfiguration.md)


</td><td>


</td></tr>
<tr><td>

[SecondaryIndexIrShape](./data-schema-types.secondaryindexirshape.md)


</td><td>

SecondaryIndex index types and query methods


</td></tr>
<tr><td>

[SelectionSet](./data-schema-types.selectionset.md)


</td><td>


</td></tr>
<tr><td>

[SetTypeSubArg](./data-schema-types.settypesubarg.md)


</td><td>

Replaces the value of a key in a complex generic type param


</td></tr>
<tr><td>

[SingularReturnValue](./data-schema-types.singularreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[SqlStatementFolderEntry](./data-schema-types.sqlstatementfolderentry.md)


</td><td>


</td></tr>
<tr><td>

[UnionToIntersection](./data-schema-types.uniontointersection.md)


</td><td>


</td></tr>
<tr><td>

[UnwrapArray](./data-schema-types.unwraparray.md)


</td><td>


</td></tr>
</tbody></table>
