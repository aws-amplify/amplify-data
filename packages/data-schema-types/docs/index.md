<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md)

## API Reference

## Packages

|  Package | Description |
|  --- | --- |
|  [@aws-amplify/data-schema-types](./data-schema-types.md) |  |

