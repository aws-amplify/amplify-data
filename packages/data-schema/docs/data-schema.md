<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema](./data-schema.md)

## data-schema package

## Enumerations

<table><thead><tr><th>

Enumeration


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[ModelRelationshipTypes](./data-schema.modelrelationshiptypes.md)


</td><td>

Model relationship types


</td></tr>
</tbody></table>

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[EnumType](./data-schema.enumtype.md)


</td><td>

Enum type definition content


</td></tr>
</tbody></table>

## Namespaces

<table><thead><tr><th>

Namespace


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[a](./data-schema.a.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Authorization](./data-schema.authorization.md)


</td><td>

Container for authorization schema definition content.


</td></tr>
<tr><td>

[ClientSchema](./data-schema.clientschema.md)


</td><td>


</td></tr>
<tr><td>

[CombinedModelSchema](./data-schema.combinedmodelschema.md)


</td><td>

CombinedModel schema definition interface


</td></tr>
<tr><td>

[CustomOperation](./data-schema.customoperation.md)


</td><td>

Custom operation definition interface


</td></tr>
<tr><td>

[CustomType](./data-schema.customtype.md)


</td><td>

Custom type container


</td></tr>
<tr><td>

[Json](./data-schema.json.md)


</td><td>

A precise, recursive Json type blows the type calculation stack without installing explicit `Json extends T ? short-circuit : ...` type checks all over the place. We may take that on later. But, this is a good-enough approximation for now.


</td></tr>
<tr><td>

[ModelDefaultIdentifier](./data-schema.modeldefaultidentifier.md)


</td><td>

Model default identifier


</td></tr>
<tr><td>

[ModelField](./data-schema.modelfield.md)


</td><td>

Public API for the chainable builder methods exposed by Model Field. The type is narrowing e.g., after calling .array() it will be omitted from intellisense suggestions


</td></tr>
<tr><td>

[ModelRelationshipField](./data-schema.modelrelationshipfield.md)


</td><td>

Model relationship field definition interface


</td></tr>
<tr><td>

[ModelRelationshipTypeArgFactory](./data-schema.modelrelationshiptypeargfactory.md)


</td><td>

Model relationship type definition content


</td></tr>
<tr><td>

[ModelSchema](./data-schema.modelschema.md)


</td><td>

Model schema definition interface


</td></tr>
<tr><td>

[ModelType](./data-schema.modeltype.md)


</td><td>

Model type definition interface


</td></tr>
<tr><td>

[Nullable](./data-schema.nullable.md)


</td><td>

Field type arg mutators


</td></tr>
<tr><td>

[RDSModelSchema](./data-schema.rdsmodelschema.md)


</td><td>

RDSModel schema definition interface


</td></tr>
<tr><td>

[RefType](./data-schema.reftype.md)


</td><td>

Reference type definition interface


</td></tr>
<tr><td>

[ResourceAuthorization](./data-schema.resourceauthorization.md)


</td><td>

Container for representing the authorization of function resources


</td></tr>
</tbody></table>
