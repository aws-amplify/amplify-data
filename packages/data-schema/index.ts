import * as a from './src';
import { ClientSchema } from './src/ClientSchema';

export { a };

export type { ClientSchema };

export * from './src/runtime/client';
export {
  addSchemaToClient,
  addSchemaToClientWithInstance,
} from './src/runtime';
