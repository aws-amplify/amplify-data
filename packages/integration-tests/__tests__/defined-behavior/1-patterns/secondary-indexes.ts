import { a, ClientSchema } from '@aws-amplify/data-schema';
import { Amplify } from 'aws-amplify';
import { Expect, Equal } from '@aws-amplify/data-schema-types';
import {
  buildAmplifyConfig,
  mockedGenerateClient,
  optionsAndHeaders,
  expectGraphqlMatches,
} from '../../utils';

describe('Custom secondary indexes', () => {
  // https://docs.amplify.aws/react/build-a-backend/data/data-modeling/secondary-index/

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('main example', () => {
    // #region covers 22ba8c0684be1400
    const schema = a.schema({
      Customer: a
        .model({
          name: a.string(),
          phoneNumber: a.phone(),
          accountRepresentativeId: a.id().required(),
        })
        .secondaryIndexes((index) => [index('accountRepresentativeId')])
        .authorization((allow) => [allow.publicApiKey()]),
    });
    // #endregion

    type Schema = ClientSchema<typeof schema>;

    test(`can retrieve records using the custom index`, async () => {
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            listCustomerByAccountRepresentativeId: {
              items: [
                {
                  name: 'name value',
                  phoneNumber: '555-555-5555',
                  accountRepresentativeId: 'one',
                },
              ],
            },
          },
        },
        {
          data: null,
          errors: [
            {
            path: null,
            locations: []
            } as any
          ]
          },
      ]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);

      // #region covers 00052f6a61b4a992
      const client = generateClient<Schema>();

      const { data, errors } =
        await client.models.Customer.listCustomerByAccountRepresentativeId({
          accountRepresentativeId: 'YOUR_REP_ID',
        });
      

      // #endregion

      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(data).toEqual(
        expect.objectContaining([
          {
            name: 'name value',
            phoneNumber: '555-555-5555',
            accountRepresentativeId: 'one',
          },
        ]),
      );

      // Ensuring `data: null` does not throw exception
      // https://github.com/aws-amplify/amplify-js/issues/13941
      expect(async () => {
        const { data: data2, errors: errors2 } =
          await client.models.Customer.listCustomerByAccountRepresentativeId({
            accountRepresentativeId: 'YOUR_REP_ID',
          });
      }).not.toThrow();
    });
    
  });

  describe('Add sort keys to secondary indexes', () => {
    // #region covers 956e34083f456707
    const schema = a.schema({
      Customer: a
        .model({
          name: a.string(),
          phoneNumber: a.phone(),
          accountRepresentativeId: a.id().required(),
        })
        .secondaryIndexes((index) => [
          index('accountRepresentativeId').sortKeys(['name']),
        ])
        .authorization((allow) => [allow.owner()]),
    });
    // #endregion

    type Schema = ClientSchema<typeof schema>;

    test(`can retrieve records using the custom index`, async () => {
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            listCustomerByAccountRepresentativeIdAndName: {
              items: [
                {
                  name: `Rene's name`,
                  phoneNumber: '555-555-5555',
                  accountRepresentativeId: 'one',
                },
              ],
            },
          },
        },
      ]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);

      // #region covers d839094b3650525c
      const client = generateClient<Schema>();

      const { data, errors } =
        await client.models.Customer.listCustomerByAccountRepresentativeIdAndName(
          {
            accountRepresentativeId: 'YOUR_REP_ID',
            name: {
              beginsWith: 'Rene',
            },
          },
        );

      // #endregion

      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(data).toEqual(
        expect.objectContaining([
          {
            name: `Rene's name`,
            phoneNumber: '555-555-5555',
            accountRepresentativeId: 'one',
          },
        ]),
      );
    });
  });

  describe('Customize the query field for secondary indexes', () => {
    // #region covers 1b750f5a650f2978
    const schema = a.schema({
      Customer: a
        .model({
          name: a.string(),
          phoneNumber: a.phone(),
          accountRepresentativeId: a.id().required(),
        })
        .secondaryIndexes((index) => [
          index('accountRepresentativeId').queryField('listByRep'),
        ])
        .authorization((allow) => [allow.owner()]),
    });

    // #endregion

    type Schema = ClientSchema<typeof schema>;

    test(`can retrieve records using the custom index`, async () => {
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: {
            listCustomerByAccountRepresentativeIdAndName: {
              items: [
                {
                  name: `Rene's name`,
                  phoneNumber: '555-555-5555',
                  accountRepresentativeId: 'one',
                },
              ],
            },
          },
        },
      ]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);

      // #region covers 4df128e1af50d457
      const client = generateClient<Schema>();

      const { data, errors } = await client.models.Customer.listByRep({
        accountRepresentativeId: 'YOUR_REP_ID',
      });
      // #endregion

      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(data).toEqual(
        expect.objectContaining([
          {
            name: `Rene's name`,
            phoneNumber: '555-555-5555',
            accountRepresentativeId: 'one',
          },
        ]),
      );
    });
  });

  describe('Customize the name of secondary indexes', () => {
    test(`generated graphql schema uses the provided name`, async () => {
      // #region covers 295dee0505f4abd0
      const schema = a.schema({
        Customer: a
          .model({
            name: a.string(),
            phoneNumber: a.phone(),
            accountRepresentativeId: a.id().required(),
          })
          .secondaryIndexes((index) => [
            index('accountRepresentativeId').name('MyCustomIndexName'),
          ])
          .authorization((allow) => [allow.owner()]),
      });

      expectGraphqlMatches(
        schema.transform().schema,
        `type Customer @model @auth(rules: [{allow: owner, ownerField: "owner"}])
        {
          name: String
          phoneNumber: AWSPhone
          accountRepresentativeId: ID! @index(name: "MyCustomIndexName", queryField: "listCustomerByAccountRepresentativeId")
        }`,
      );
      // #endregion
    });
  });
});
