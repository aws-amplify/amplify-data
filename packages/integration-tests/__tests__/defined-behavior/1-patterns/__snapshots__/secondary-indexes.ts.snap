// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom secondary indexes Add sort keys to secondary indexes can retrieve records using the custom index 1`] = `
[
  [
    {
      "apiKey": undefined,
      "authMode": undefined,
      "authToken": undefined,
      "endpoint": undefined,
      "query": "query($accountRepresentativeId: ID!,$name: ModelStringKeyConditionInput,$filter: ModelCustomerFilterInput,$sortDirection: ModelSortDirection,$limit: Int,$nextToken: String) { listCustomerByAccountRepresentativeIdAndName(accountRepresentativeId: $accountRepresentativeId,name: $name,filter: $filter,sortDirection: $sortDirection,limit: $limit,nextToken: $nextToken) { items { id name phoneNumber accountRepresentativeId createdAt updatedAt owner } nextToken __typename } }",
      "variables": {
        "accountRepresentativeId": "YOUR_REP_ID",
        "name": {
          "beginsWith": "Rene",
        },
      },
    },
    {},
  ],
]
`;

exports[`Custom secondary indexes Customize the query field for secondary indexes can retrieve records using the custom index 1`] = `
[
  [
    {
      "apiKey": undefined,
      "authMode": undefined,
      "authToken": undefined,
      "endpoint": undefined,
      "query": "query($accountRepresentativeId: ID!,$filter: ModelCustomerFilterInput,$sortDirection: ModelSortDirection,$limit: Int,$nextToken: String) { listByRep(accountRepresentativeId: $accountRepresentativeId,filter: $filter,sortDirection: $sortDirection,limit: $limit,nextToken: $nextToken) { items { id name phoneNumber accountRepresentativeId createdAt updatedAt owner } nextToken __typename } }",
      "variables": {
        "accountRepresentativeId": "YOUR_REP_ID",
      },
    },
    {},
  ],
]
`;

exports[`Custom secondary indexes main example can retrieve records using the custom index 1`] = `
[
  [
    {
      "apiKey": undefined,
      "authMode": undefined,
      "authToken": undefined,
      "endpoint": undefined,
      "query": "query($accountRepresentativeId: ID!,$filter: ModelCustomerFilterInput,$sortDirection: ModelSortDirection,$limit: Int,$nextToken: String) { listCustomerByAccountRepresentativeId(accountRepresentativeId: $accountRepresentativeId,filter: $filter,sortDirection: $sortDirection,limit: $limit,nextToken: $nextToken) { items { id name phoneNumber accountRepresentativeId createdAt updatedAt } nextToken __typename } }",
      "variables": {
        "accountRepresentativeId": "YOUR_REP_ID",
      },
    },
    {},
  ],
]
`;
